lib/AST/ASTPrinter.cpp

```diff
@@ -2750,8 +2751,13 @@ void PrintAST::printEnumElement(EnumElementDecl *elt) {
                                       ->castTo<AnyFunctionType>()
                                       ->getParams();
     }
+
+    // @escaping is not valid in enum element position, even though the
+    // attribute is implicitly added. Ignore it when printing the parameters.
+    Options.ExcludeAttrList.push_back(TAK_escaping);
     printParameterList(PL, params,
```

Comment:
Would be cleaner to add a flag to printParameterList() for this...

Reply:
I started with that, then switched to this because there are a bunch of places where we put attrs into this list and remove them after we're done (`printAttributes` being one of them). I don't like the idea that someone could add something else to the list and then unbalance this, though.

Reply:
ok but maybe those other places should be cleaned up instead :)

If you must add/remove attrs like this, change the ExclueAttrsList into a set and assert that removing the element returns success.

Reply:
I played around a bit with that a while ago, but there's this faux-scoping thing that happens inside `printAttributes` where we add several attributes and remove them all in one go. It'd be a bit more work but yeah, I think this could be rethought to be a bit safer. Maybe like a scoped counter or something that keeps a map of the number of times something's inserted? We want to be able to track duplicates, because otherwise everyone would have to check if the value is in the set already before inserting. I'll clean this up tomorrow.